docs/specifications/tests/DNSSEC-TP/dnssec02.md

```diff
@@ -1,69 +1,159 @@
-## DNSSEC02: DS must match a DNSKEY in the designated zone
+# DNSSEC02: DS must match a valid DNSKEY in the child zone
 
-### Test case identifier
-**DNSSEC02** DS must match a DNSKEY in the designated zone
+## Test case identifier
+**DNSSEC02**
 
-### Objective
+## Objective
 
 DNS delegations from a parent to a child are secured with DNSSEC by
 publishing one or several Delgation Signer (DS) records in the parent
-zone, along with the NS records for the delegation. For the secure
-delegation to work, the DS record must match at least one DNSKEY record
-in the child zone.
+zone, along with the NS records for the delegation.
+
+For the secure delegation to work, at least one DS record must match a
+DNSKEY record in the child zone ([RFC 4035][RFC 4035#5], section 5).
+Each DS record should match a DNSKEY record in the child zone. More
+than one DS may match the same DNSKEY. The DNSKEY that the DS record
+refer to must be used to sign the DNSKEY RRset in the child zone
+([RFC 4035][RFC 4035#5], section 5).
+
+The DNSKEY record that the DS record refer to must have bit 7
+("Zone Key flag") set in the DNSKEY RR Flags ([RFC 4034][RFC 4034#5.2], 
+section 5.2).
+
+Bit 15 ("Secure Entry Point flag") on a DNSKEY record signals that it
+is meant to be a KSK and pointed out by a DS record. It is noted if
+the DNSKEY record that the DS points at does not have that flag set
+([RFC 4034][RFC 4034#2.1.1], section 2.1.1).
+
+## Inputs
+
+* "Child Zone" - The domain name to be tested.
+* "Test Type" - The test type with value "undelegated" or "normal".
+* "Undelegated DS" - The DS record or records submitted
+  (only if *Test Type* is undelegated).
+
+## Ordered description of steps to be taken to execute the test case
+
+1. Create an empty set, "DS records".
+
+2. If the *Test Type* is "undelegated, then:
+   1. Add *Undelegated DS* set to *DS records*.
+
+3. Else, do:
+   1. Create a DS query with DO flag set for the name of the
+      *Child Zone* (*Test Type* is "normal").
+   2. Retrieve all name server IP addresses for the parent zone of
+      *Child Zone* using [Method1] ("Parent NS IP").
+   3. For each IP address in *Parent NS IP* do:
+      1. Send the DS query over UDP and collect the response.
+      2. If there is no DNS response, then output *[NO_RESPONSE]*.
+      3. Else, if the RCODE is not NOERROR, then output
+         *[UNEXPECTED_RESPONSE_DS]*.
+      4. Else, go to next IP address if there is no DS in the
+         response.
+      5. Else, extract all DS records from the DNS response and
+         add them to *DS Records*.
+
+4. If *DS Records* is empty, exit this test case.
```

Comment:
Shouldn't the test case exit with an error message if DS records are empty

Reply:
No, because it is not an error not to have a DS. The purpose of this test case is to verify that DS, if any, matches a valid DNSKEY.